readme.txt
JEQ
=====================================================
version 1.01, September 1, 2005

Description
-----------
JEQ is class library written in pure Java. It implements the infinite impulse
response (IIR) algorithm for bandpass filtering of sound stream. So, it
represents a backend code for stereo 10/15/25/31-band equalizer.

The main distributables is:

1. jeq.jar - limited version (equalizer).
2. jeqspi.jar - full version (equalizer + Java Sound API support).

This is a non-commercial project and anyone can add his contribution.
JEQ is distributed under the terms of the FSF Lesser Gnu Public License (see lgpl.txt).

Instructions
------------
Unzip to installation directory.

Required libraries are distributed in the "distrib" directory.
Web Site: http://jeq.sourceforge.net/

FAQ : 
------------
- How to test the JEQ quickly?

  Please, look the bin directory for examples.

- How to use the JEQ?

  Please, look the davaguine.jmac.test package sources for working examples.

- How much CPU JMAC needs to run?

  less than 6% - AMD64/WinXP+J2SE 1.5.0
  
- How to contact JEQ developers?
  Please, visit project web page - http://jeq.sourceforge.net or you can email
  directly to davagin@mail.ru.


changelog.txt
JEQ Changelog
===================

version 1.0 (September 1, 2005)
-------------
Initial version

version 1.01 (September 11, 2005)
-------------
Added improvements to equalizer controls

src/davaguine/jeq/test/JavaSoundSimpleAudioPlayer.java
/*
 *  21.04.2004 Original verion. davagin@udm.ru.
 *-----------------------------------------------------------------------
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *----------------------------------------------------------------------
 */

package davaguine.jeq.test;

import davaguine.jeq.spi.EqualizerInputStream;

import javax.sound.sampled.*;
import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;

/**
 * Author: Dmitry Vaguine
 * Date: 12.03.2004
 * Time: 13:35:13
 */
public class JavaSoundSimpleAudioPlayer {
    /**
     * Plays audio from given file names.
     */
    public static void main(String[] args) {
        // Check for given sound file names.
        if (args.length < 1) {
            System.out.println("Play usage:");
            System.out.println("\tjava Play <sound file names>*");
            System.exit(0);
        }

        // Process arguments.
        for (int i = 0; i < args.length; i++)
            playAudioFile(args[i]);

        // Must exit explicitly since audio creates non-daemon threads.
        System.exit(0);
    } // main

    /**
     * Play audio from the given file name.
     */
    public static void playAudioFile(String fileName) {
        try {
            AudioInputStream audioInputStream;
            try {
                URL url = new URL(fileName);
                audioInputStream = AudioSystem.getAudioInputStream(url);
            } catch (MalformedURLException e) {
                File soundFile = new File(fileName);
                audioInputStream = AudioSystem.getAudioInputStream(soundFile);
            }

            // Create a stream from the given file.
            // Throws IOException or UnsupportedAudioFileException
            playAudioStream(audioInputStream);
        } catch (Exception e) {
            System.out.println("Problem with file " + fileName + ":");
            e.printStackTrace();
        }
    } // playAudioFile

    /**
     * Plays audio from the given audio input stream.
     */
    public static void playAudioStream(AudioInputStream audioInputStream) {
        // Audio format provides information like sample rate, size, channels.
        AudioFormat audioFormat = audioInputStream.getFormat();

        // Convert compressed audio data to uncompressed PCM format.
        if (audioFormat.getEncoding() != AudioFormat.Encoding.PCM_SIGNED) {
            AudioFormat newFormat = new AudioFormat(audioFormat.getSampleRate(),
                    audioFormat.getSampleSizeInBits(),
                    audioFormat.getChannels(),
                    true,
                    false);
            System.out.println("Converting audio format to " + newFormat);
            AudioInputStream newStream = AudioSystem.getAudioInputStream(newFormat, audioInputStream);
            audioFormat = newFormat;
            audioInputStream = newStream;
        }

        // Open a data line to play our type of sampled audio.
        // Use SourceDataLine for play and TargetDataLine for record.
        DataLine.Info info = new DataLine.Info(SourceDataLine.class, audioFormat);
        if (!AudioSystem.isLineSupported(info)) {
            System.out.println("Play.playAudioStream does not handle this type of audio on this system.");
            return;
        }

        try {
            // Create a SourceDataLine for play back (throws LineUnavailableException).
            SourceDataLine dataLine = (SourceDataLine) AudioSystem.getLine(info);
            System.out.println("SourceDataLine class=" + dataLine.getClass());

            // The line acquires system resources (throws LineAvailableException).
            dataLine.open(audioFormat);

            // Adjust the volume on the output line.
            if (dataLine.isControlSupported(FloatControl.Type.MASTER_GAIN)) {
                FloatControl volume = (FloatControl) dataLine.getControl(FloatControl.Type.MASTER_GAIN);
                volume.setValue((volume.getMaximum() - volume.getMinimum()) * 0.7f + volume.getMinimum());
            }

            // Allows the line to move data in and out to a port.
            dataLine.start();

            EqualizerInputStream eq = new EqualizerInputStream(audioInputStream, 31);

            // Create a buffer for moving data from the audio stream to the line.
            int bufferSize = (int) audioFormat.getSampleRate() * audioFormat.getFrameSize();
            byte[] buffer = new byte[bufferSize];

            // Move the data until done or there is an error.
            try {
                int bytesRead = 0;
                while (bytesRead >= 0) {
                    bytesRead = eq.read(buffer, 0, buffer.length);
                    if (bytesRead >= 0)
                        dataLine.write(buffer, 0, bytesRead);
                } // while
            } catch (IOException e) {
                e.printStackTrace();
            }

            System.out.println("Play.playAudioStream draining line.");
            // Continues data line I/O until its buffer is drained.
            dataLine.drain();

            System.out.println("Play.playAudioStream closing line.");
            // Closes the data line, freeing any resources such as the audio device.
            dataLine.close();
        } catch (LineUnavailableException e) {
            e.printStackTrace();
        }
    } // playAudioStream
}


src/davaguine/jeq/core/XYData.java
/*
 *  21.04.2004 Original verion. davagin@udm.ru.
 *-----------------------------------------------------------------------
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *----------------------------------------------------------------------
 */
package davaguine.jeq.core;

/**
 * Structure for storing XYData of equalizer.
 * Author: Dmitry Vaguine
 * Date: 02.05.2004
 * Time: 12:00:29
 */
public class XYData {
    /**
     * X data
     */
    public double x[] = new double[3]; /* x[n], x[n-1], x[n-2] */
    /**
     * Y data
     */
    public double y[] = new double[3]; /* y[n], y[n-1], y[n-2] */

    /**
     * Constructs new XYData object
     */
    public XYData() {
        zero();
    }

    /**
     * Fills all content with zero
     */
    public void zero() {
        for (int i = 0; i < 3; i++) {
            x[i] = 0;
            y[i] = 0;
        }
    }
}


src/davaguine/jeq/core/IIRControls.java
/*
 *  21.04.2004 Original verion. davagin@udm.ru.
 *-----------------------------------------------------------------------
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *----------------------------------------------------------------------
 */
package davaguine.jeq.core;

/**
 * Constols of equalizer
 * Author: Dmitry Vaguine
 * Date: 02.05.2004
 * Time: 12:00:29
 */
public class IIRControls {
    /**
     * Volume gain
     * values should be between 0.0 and 1.0
     */
    private float preamp[];
    /**
     * Gain for each band
     * values should be between -0.2 and 1.0
     */
    private float bands[][];

    /**
     * Creates new IIRControls object for given number of bands
     *
     * @param bandsnum is the number of bands
     * @param channels is the number of channels
     */
    public IIRControls(int bandsnum, int channels) {
        preamp = new float[channels];
        bands = new float[bandsnum][channels];
        for (int j = 0; j < channels; j++) {
            preamp[j] = 1.0f;
            for (int i = 0; i < bandsnum; i++)
                bands[i][j] = 0f;
        }
    }

    /**
     * Returns the maximum value for band control
     *
     * @return the maximum value for band control
     */
    public float getMaximumBandValue() {
        return 1.0f;
    }

    /**
     * Returns the minimum value for band control
     *
     * @return the minimum value for band control
     */
    public float getMinimumBandValue() {
        return -0.2f;
    }

    /**
     * Returns the maximum value for band control (in Db)
     *
     * @return the maximum value for band control
     */
    public float getMaximumBandDbValue() {
        return 12;
    }

    /**
     * Returns the minimum value for band control (in Db)
     *
     * @return the minimum value for band control
     */
    public float getMinimumBandDbValue() {
        return -12f;
    }

    /**
     * Returns the maximum value for preamp control
     *
     * @return the maximum value for preamp control
     */
    public float getMaximumPreampValue() {
        return 1.0f;
    }

    /**
     * Returns the minimum value for preamp control
     *
     * @return the minimum value for preamp control
     */
    public float getMinimumPreampValue() {
        return 0f;
    }

    /**
     * Returns the maximum value for preamp control (in Db)
     *
     * @return the maximum value for preamp control
     */
    public float getMaximumPreampDbValue() {
        return 12f;
    }

    /**
     * Returns the minimum value for preamp control (in Db)
     *
     * @return the minimum value for preamp control
     */
    public float getMinimumPreampDbValue() {
        return -12f;
    }

    /**
     * Returns bands array
     *
     * @return bands array
     */
    float[][] getBands() {
        return bands;
    }

    /**
     * Returns preamp array
     *
     * @return preamp array
     */
    float[] getPreamp() {
        return preamp;
    }

    /**
     * Returns value of control for given band and channel
     *
     * @param band    is the index of band
     * @param channel is the index of channel
     * @return the value
     */
    public float getBandValue(int band, int channel) {
        return bands[band][channel];
    }

    /**
     * Setter for value of control for given band and channel
     *
     * @param band    is the index of band
     * @param channel is the index of channel
     * @param value   is the new value
     */
    public void setBandValue(int band, int channel, float value) {
        bands[band][channel] = value;
    }

    /**
     * Setter for value of control for given band and channel (in Db)
     *
     * @param band    is the index of band
     * @param channel is the index of channel
     * @param value   is the new value
     */
    void setBandDbValue(int band, int channel, float value) {
        /* Map the gain and preamp values */
        /* -12dB .. 12dB mapping */
        bands[band][channel] = (float) (2.5220207857061455181125E-01 *
                Math.exp(8.0178361802353992349168E-02 * value)
                - 2.5220207852836562523180E-01);
    }

    /**
     * Returns value of preamp control for given channel
     *
     * @param channel is the index of channel
     * @return the value
     */
    public float getPreampValue(int channel) {
        return preamp[channel];
    }

    /**
     * Setter for value of preamp control for given channel
     *
     * @param channel is the index of channel
     * @param value   is the new value
     */
    public void setPreampValue(int channel, float value) {
        preamp[channel] = value;
    }

    /**
     * Setter for value of preamp control for given channel (in Db)
     *
     * @param channel is the index of channel
     * @param value   is the new value
     */
    public void setPreampDbValue(int channel, float value) {
        /* -12dB .. 12dB mapping */
        preamp[channel] = (float) (9.9999946497217584440165E-01 *
                Math.exp(6.9314738656671842642609E-02 * value)
                + 3.7119444716771825623636E-07);
    }
}


src/davaguine/jeq/core/IIRCoefficients.java
/*
 *  21.04.2004 Original verion. davagin@udm.ru.
 *-----------------------------------------------------------------------
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *----------------------------------------------------------------------
 */
package davaguine.jeq.core;

/**
 * Author: Dmitry Vaguine
 * Date: 02.05.2004
 * Time: 12:00:29
 */
public class IIRCoefficients {
    public double beta;
    public double alpha;
    public double gamma;

    public IIRCoefficients(double beta, double alpha, double gamma) {
        this.beta = beta;
        this.alpha = alpha;
        this.gamma = gamma;
    }
}


src/davaguine/jeq/core/IIR.java
/*
 *  21.04.2004 Original verion. davagin@udm.ru.
 *-----------------------------------------------------------------------
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *----------------------------------------------------------------------
 */
package davaguine.jeq.core;

/**
 * Generic wrapper around IIR algorithm.
 * Author: Dmitry Vaguine
 * Date: 02.05.2004
 * Time: 12:00:29
 */
public class IIR extends IIRBase {
    /**
     * Max number of channels supported
     */
    public final static int EQ_MAX_CHANNELS = 2;
    /**
     * Max bands supported by the code
     */
    public final static int EQ_MAX_BANDS = 31;
    /**
     * Supported sample rates
     */
    public final static float EQ_11025_RATE = 11025;
    public final static float EQ_22050_RATE = 22050;
    public final static float EQ_44100_RATE = 44100;
    public final static float EQ_48000_RATE = 48000;
    /**
     * Supported number of bands
     */
    public final static int EQ_10_BANDS = 10;
    public final static int EQ_15_BANDS = 15;
    public final static int EQ_25_BANDS = 25;
    public final static int EQ_31_BANDS = 31;

    /* Indexes for the history arrays
     * These have to be kept between calls to this function
     * hence they are static */
    private int i;
    private int j;
    private int k;

    /* History for two filters */
    private XYData[][] dataHistory = new XYData[EQ_MAX_BANDS][EQ_MAX_CHANNELS];
    private XYData[][] dataHistory2 = new XYData[EQ_MAX_BANDS][EQ_MAX_CHANNELS];

    /* Coefficients */
    private IIRCoefficients[] iircf;

    /* Equalizer config */
    private IIRControls eqcfg;
    /* rate */
    private float rate;
    /* channels */
    private int channels;
    /* bands */
    private int bands;

    /**
     * Constructs equalizer with given config
     *
     * @param bands    is the number of bands to be used
     * @param rate     is the sample rate of equalizer
     * @param channels is the number of channels
     */
    public IIR(int bands, float rate, int channels) {
        this.rate = rate;
        this.channels = channels;
        this.bands = bands;
        this.eqcfg = new IIRControls(bands, channels);

        if (!isParamsSupported(bands, rate, channels))
            throw new IllegalArgumentException("Unsupported parameters");

        initIIR();
    }

    /**
     * Returns Controls of equalizer
     *
     * @return Controls of equalizer
     */
    public IIRControls getControls() {
        return eqcfg;
    }

    /**
     * This is special method for checking of supported parameters of equalizer
     *
     * @param bands    is the number of bands
     * @param rate     is the sample rate of data
     * @param channels is the number of channels
     * @return true if parameters are supported
     */
    public static boolean isParamsSupported(int bands, float rate, int channels) {
        if (rate != EQ_11025_RATE && rate != EQ_22050_RATE && rate != EQ_44100_RATE && rate != EQ_48000_RATE)
            return false;

        switch (bands) {
            case EQ_10_BANDS:
            case EQ_15_BANDS:
            case EQ_25_BANDS:
            case EQ_31_BANDS:
                break;
            default:
                return false;
        }

        switch (channels) {
            case 1:
            case 2:
                break;
            default:
                return false;
        }

        return (rate != EQ_11025_RATE && rate != EQ_22050_RATE) || bands == EQ_10_BANDS;
    }

    /* Init the filters */
    private void initIIR() {
        setFilters();
        for (int ii = 0; ii < EQ_MAX_BANDS; ii++)
            for (int jj = 0; jj < EQ_MAX_CHANNELS; jj++) {
                dataHistory[ii][jj] = new XYData();
                dataHistory2[ii][jj] = new XYData();
            }
        i = 0;
        j = 2;
        k = 1;
    }

    private void setFilters() {
        if (rate == EQ_11025_RATE)
            iircf = iir_cf10_11k_11025;
        else if (rate == EQ_22050_RATE)
            iircf = iir_cf10_22k_22050;
        else if (rate == EQ_44100_RATE) {
            switch (bands) {
                case 31:
                    iircf = iir_cf31_44100;
                    break;
                case 25:
                    iircf = iir_cf25_44100;
                    break;
                case 15:
                    iircf = iir_cf15_44100;
                    break;
                default:
                    iircf = iir_cf10_44100;
                    break;
            }
        } else if (rate == EQ_48000_RATE) {
            switch (bands) {
                case 31:
                    iircf = iir_cf31_48000;
                    break;
                case 25:
                    iircf = iir_cf25_48000;
                    break;
                case 15:
                    iircf = iir_cf15_48000;
                    break;
                default:
                    iircf = iir_cf10_48000;
                    break;
            }
        }
    }

    /**
     * Clear filter history.
     */
    public void cleanHistory() {
        /* Zero the history arrays */
        for (int ii = 0; ii < EQ_MAX_BANDS; ii++)
            for (int jj = 0; jj < EQ_MAX_CHANNELS; jj++) {
                dataHistory[ii][jj].zero();
                dataHistory2[ii][jj].zero();
            }
        i = 0;
        j = 2;
        k = 1;
    }

    /**
     * Main filtering method.
     *
     * @param data   - data to be filtered
     * @param length - length of data in buffer
     */
    public void iir(int[] data, int length) {
        int index, band, channel;
        float eqpreamp[] = eqcfg.getPreamp();
        float eqbands[][] = eqcfg.getBands();
        double pcm, out;

        /**
         * IIR filter equation is
         * y[n] = 2 * (alpha*(x[n]-x[n-2]) + gamma*y[n-1] - beta*y[n-2])
         *
         * NOTE: The 2 factor was introduced in the coefficients to save
         * 			a multiplication
         *
         * This algorithm cascades two filters to get nice filtering
         * at the expense of extra CPU cycles
         */
        IIRCoefficients tempcf;
        XYData tempd;
        for (index = 0; index < length; index += channels) {
            /* For each channel */
            for (channel = 0; channel < channels; channel++) {
                /* Preamp gain */
                pcm = data[index + channel] * eqpreamp[channel];

                out = 0f;
                /* For each band */
                for (band = 0; band < bands; band++) {
                    /* Store Xi(n) */
                    tempd = dataHistory[band][channel];
                    tempd.x[i] = pcm;
                    /* Calculate and store Yi(n) */
                    tempcf = iircf[band];
                    tempd.y[i] =
                            (
                                    /* 		= alpha * [x(n)-x(n-2)] */
                                    tempcf.alpha * (pcm - tempd.x[k])
                                            /* 		+ gamma * y(n-1) */
                                            + tempcf.gamma * tempd.y[j]
                                            /* 		- beta * y(n-2) */
                                            - tempcf.beta * tempd.y[k]
                            );
                    /*
                     * The multiplication by 2.0 was 'moved' into the coefficients to save
                     * CPU cycles here */
                    /* Apply the gain  */
                    out += (tempd.y[i] * eqbands[band][channel]); // * 2.0;
                } /* For each band */

                /* Volume stuff
                   Scale down original PCM sample and add it to the filters
                   output. This substitutes the multiplication by 0.25
                   Go back to use the floating point multiplication before the
                   conversion to give more dynamic range
                   */
                out += (pcm * 0.25);

                /* Normalize the output */
                out *= 4;

                /* Round and convert to integer */
                data[index + channel] = (int) out;
            } /* For each channel */

            i++;
            j++;
            k++;

            /* Wrap around the indexes */
            if (i == 3)
                i = 0;
            else if (j == 3)
                j = 0;
            else
                k = 0;


        }/* For each pair of samples */
    }
}
